File: elixir/sublist/lib/sublist.ex
defmodule Sublist do
  @doc """
  Returns whether the first list is a sublist or a superlist of the second list
  and if not whether it is equal or unequal to the second list.
  """
  def compare(a, b) do
    cond do
      a === b ->
        :equal

      sublist?(a, b) ->
        :sublist

      sublist?(b, a) ->
        :superlist

      true
        -> :unequal
    end
  end

  defp sublist?(small_list, _) when length(small_list) == 0, do: true
  defp sublist?(small_list, big_list) do
    big_list
    |> Enum.chunk_every(length(small_list), 1, :discard)
    |> Enum.any?(& &1 === small_list)
  end
end

File: clojure/sublist/src/sublist.clj
(ns sublist
  (:use [clojure.core]))

(defn subseq?
  ;; ([s1 s2] (subseq? s1 s2 false))
  ;; ([s1 s2 partial-match]
  ;;  (cond (empty? s1) true
  ;;        ;;
  ;;        (empty? s2) false
  ;;        ;;
  ;;        (= (first s1) (first s2))
  ;;        (subseq? (rest s1) (rest s2) true)
  ;;        ;;
  ;;        :else (if partial-match
  ;;                false
  ;;                (subseq? s1 (rest s2) false))))
  [s1 s2]
  (loop [s1 s1
         s2 s2
         ls1 (count s1)]
    (and (<= ls1 (count s2))
         (or (= s1 (take ls1 s2))
             (recur s1 (rest s2) ls1)))))

(defn classify [s1 s2]
  (cond (= s1 s2) :equal
        (subseq? s1 s2) :sublist
        (subseq? s2 s1) :superlist
        :else :unequal))
